torchgeo/datasets/oscd.py

```diff
@@ -0,0 +1,313 @@
+        self.split = split
+        self.bands = bands
+        self.transforms = transforms
+        self.download = download
+        self.checksum = checksum
+
+        self._verify()
+
+        self.files = self._load_files()
+
+    def __getitem__(self, index: int) -> Dict[str, Tensor]:
+        """Return an index within the dataset.
+
+        Args:
+            index: index to return
+
+        Returns:
+            data and label at that index
+        """
+        files = self.files[index]
+        image1 = self._load_image(files["images1"])
+        image2 = self._load_image(files["images2"])
+        mask = self._load_target(str(files["mask"]))
+
+        image = torch.stack(tensors=[image1, image2], dim=0)
+        sample = {"image": image, "mask": mask}
+
+        if self.transforms is not None:
+            sample = self.transforms(sample)
+
+        return sample
+
+    def __len__(self) -> int:
+        """Return the number of data points in the dataset.
+
+        Returns:
+            length of the dataset
+        """
+        return len(self.files)
+
+    def _load_files(self) -> List[Dict[str, Union[str, Sequence[str]]]]:
+        regions = []
+        labels_root = os.path.join(
+            self.root,
+            f"Onera Satellite Change Detection dataset - {self.split.capitalize()} "
+            + "Labels",
+        )
+        images_root = os.path.join(
+            self.root, "Onera Satellite Change Detection dataset - Images"
+        )
+        folders = glob.glob(os.path.join(labels_root, "*/"))
+        for folder in folders:
+            region = folder.split(os.sep)[-2]
+            mask = os.path.join(labels_root, region, "cm", "cm.png")
+
+            def get_image_paths(ind: int) -> List[str]:
+                return sorted(
+                    glob.glob(
+                        os.path.join(images_root, region, f"imgs_{ind}_rect", "*.tif")
+                    ),
+                    key=sort_sentinel2_bands,
+                )
+
+            images1, images2 = get_image_paths(1), get_image_paths(2)
+            if self.bands == "rgb":
+                images1, images2 = images1[1:4][::-1], images2[1:4][::-1]
+
+            with open(os.path.join(images_root, region, "dates.txt")) as f:
+                dates = tuple(
+                    [line.split()[-1] for line in f.read().strip().splitlines()]
+                )
+
+            regions.append(
+                dict(
+                    region=region,
+                    images1=images1,
+                    images2=images2,
+                    mask=mask,
+                    dates=dates,
+                )
+            )
+
+        return regions
+
+    def _load_image(self, paths: Sequence[str]) -> Tensor:
+        """Load a single image.
+
+        Args:
+            path: path to the image
+
+        Returns:
+            the image
+        """
+        images = []
+        for path in paths:
+            with rasterio.open(path) as f:
```

Comment:
I'm getting errors when using this dataset:
```python
>>> OSCD()[0]
/anaconda/envs/torchgeo/lib/python3.9/site-packages/rasterio/__init__.py:220: NotGeoreferencedWarning: Dataset has no geotransform, gcps, or rpcs. The identity matrix be returned.
  s = DatasetReader(path, driver=driver, sharing=sharing, **kwargs)
```
It looks like some of the dataset files contain geospatial metadata and others don't. Can you look into this? We should use PIL to read non-geospatial files instead of rasterio.

Reply:
Just checked and the original files are multispectral geotiffs and not georeferenced. Don't think we are able to avoid it this time unless we filter the NotGeoreferencedWarning.

Reply:
Isn't each band stored as a separate file?

Reply:
You're right, we can use pillow instead here.

Reply:
What's the motivation behind using pillow over rasterio (and filtering the `NotGeoreferencedWarning`) for non-geospatial rasters?  Are there other gotchas involved?

Reply:
> Is there a preferred way of checking if a file contains geospatial metadata?

Not really, see #139. I'm hoping that things aren't completely random and that there is some pattern to which files contain geospatial metadata and which don't.

> What's the motivation behind using pillow over rasterio (and filtering the NotGeoreferencedWarning) for non-geospatial rasters? Are there other gotchas involved?

These files are not geospatial, so there's no reason to use a library like rasterio to load them. Libraries like PIL are more than sufficient and we already have a dependency on both. If we can't tell ahead of time which files will be geospatial, we could filter the warning, but this seems like more of a hack than a solution.

Reply:
> Is there a preferred way of checking if a file contains geospatial metadata?

Simply doing `src = rasterio.open(path_to_file)` will trigger the warning if the geotiff contains no crs/gcps/transform/etc.

Reply:
I think the only edge case is if the files are multispectral. Pillow can only read up to 4 channels iirc. In this case, the files are single channel so pillow can handle them.

Reply:
> > What's the motivation behind using pillow over rasterio (and filtering the NotGeoreferencedWarning) for non-geospatial rasters? Are there other gotchas involved?
> 
> These files are not geospatial, so there's no reason to use a library like rasterio to load them. Libraries like PIL are more than sufficient and we already have a dependency on both. If we can't tell ahead of time which files will be geospatial, we could filter the warning, but this seems like more of a hack than a solution.

I guess I'm wondering why PIL is a dependency at all when rasterio can read any raster (geospatial or not) albeit with the `NotGeoreferenced` hack previously mentioned. If we get a non-georeferenced multispectral (>3 channels) non `uint8` images PIL can't be used to read those files and we'll have to rely on rasterio right? 

To be fair pillow is aware of this and is tracking this python-pillow/Pillow#1888 but it's been a while.

Reply:
`rasterio` can read any raster depending on which drivers GDAL is built with, although the same is true for `pillow`. I think `pillow` is definitely a more popular library with more users/updates and fewer bugs, so I tend to prefer it over `rasterio` for non-geospatial files. I don't really want to have to filter out that warning everywhere we use `pillow`. Also, we have a dependency on `torchvision`, which has a dep on `pillow`, so we can't really avoid `pillow` anyway.

Reply:
That's fair. Thanks for clearing that up